clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1822,6 +1874,85 @@ static std::string getCPPTypeString(QualType Ty) {
   return eraseAnonNamespace(Ty.getAsString(P));
 }
 
+static void printArguments(raw_ostream &ArgOS, ArrayRef<TemplateArgument> Args,
+                           const PrintingPolicy &P, bool ParameterPack) {
+
+  if (!ParameterPack)
+    ArgOS << "<";
+
+  bool FirstArg = true;
+  const char *Comma = ", ";
+
+  for (unsigned I = 0; I < Args.size(); I++) {
+    const TemplateArgument &Arg = Args[I];
+    if (Arg.getKind() == TemplateArgument::ArgKind::Pack) {
+      if (Arg.pack_size() && !FirstArg)
+        ArgOS << Comma;
+      printArguments(ArgOS, Arg.getPackAsArray(), P, true);
+    } else if (Arg.getKind() == TemplateArgument::ArgKind::Integral) {
+      if (!FirstArg)
+        ArgOS << Comma;
+
+      QualType T = Arg.getIntegralType();
+      const EnumType *ET = T->getAs<EnumType>();
+
+      // Enums defined in SYCL headers are not changed to explicit values since
+      // the definition is visible in integration header.
+      if (ET && !checkIfDeclInCLNameSpace(ET->getDecl()->getDeclContext())) {
+        const llvm::APSInt &Val = Arg.getAsIntegral();
+        ArgOS << "<(" << ET->getDecl()->getQualifiedNameAsString() << ")>"
+              << Val;
+      } else {
+        Arg.print(P, ArgOS);
+      }
+    } else {
+      if (!FirstArg)
+        ArgOS << Comma;
+      LangOptions LO;
+      PrintingPolicy TypePolicy(LO);
+      TypePolicy.SuppressTypedefs = true;
+      TypePolicy.SuppressTagKeyword = true;
+      Arg.print(TypePolicy, ArgOS);
+    }
+    FirstArg = false;
+  }
+
+  if (!ParameterPack)
+    ArgOS << ">";
+}
+
+static std::string getKernelNameTypeString(QualType T) {
+
+  const CXXRecordDecl *RD = T->getAsCXXRecordDecl();
+
+  if (!RD)
+    return getCPPTypeString(T);
+
+  // If kernel name type is a template specialization with enum type
+  // template parameters, enumerators in name type string should be
+  // replaced  with their underlying value since the enum definition
+  // is not visible in integration header.
+  if (const auto *TSD = dyn_cast<ClassTemplateSpecializationDecl>(RD)) {
+    LangOptions LO;
+    PrintingPolicy P(LO);
+    P.SuppressTypedefs = true;
+    SmallString<64> Buf;
+    llvm::raw_svector_ostream ArgOS(Buf);
+
+    // Print template class name
+    TSD->printQualifiedName(ArgOS, P);
```

Comment:
This is where/how the name is currently printed. I cannot get underlying type and print it here because that prints template arguments as well.